feat: fill in contributing (#1)
aneoconsulting/ArmoniK.Community: CONTRIBUTING.md

# Contributing to ArmoniK

Hi! We are excited that you are interested in ArmoniK and want to make it even better! This document is a general guide we would like our contributors to follow. Before submitting your contribution, please make sure to take a moment and read through the following guide.

## Discuss First !

> Before attempting to make a feature pull request, open an issue and discuss it with us. It will help maintainers understand what you want to do as well as make sure that you go in a direction that is suitable for ArmoniK' future.

Reviewer: Add a paragraph about the discussion section in ArmoniK if they have a question or an idea.

## Submission guidelines

### Submit an issue

> Before submitting a new issue, please take a look at the existing [ArmoniK issues](https://github.com/aneoconsulting/ArmoniK/issues) to see if someone already mentioned your issue. If it already exists, it may inform you of workaround readily available.

Reviewer: take a look at the issue and the discussion in ArmoniK

We want to fix issues as soon as possible, but before fixing a bug, we need to reproduce and confirm it. In order to do so, we require you to provide a minimal reproduction example. This will give us important information without having to ask you multiple complementary questions.

> A minimal reproductoion scenario allows us to quickly confirm a bug, isolate it and confirm that we are really fixing it.

Reviewer: describe wha a "minimal reproduction scenario" is

> Unfortunately, we are not able to investigate or fix bugs without a minimal reproduction sample, so if we don't hear back from you, we will close the issue since it is missing information to be reproduced.

Reviewer: if user can provide more informations in after the issue was closed, we can reopened it.

Helping us to easily reproduce an issue also helps you. Our developpers will be able to more easily pinpoint the issue and provide a correction. Thus, solving your issue faster.

You can file new issues by selecting from our [new issues templates](https://github.com/aneoconsulting/ArmoniK/issues/new/choose) and filling it.

### Commit Convention

We use [Conventional Commits](https://www.conventionalcommits.org/) for commit messages. We use it to auto-generate our changelog.

> Only `fix:` and `feat:` will be presented in the changelog.

Reviewer: it's not the case. https://github.com/unjs/changelogen/blob/main/src/git.ts#L91 (every commit are in the changelog while they respect Conventional Commits).

And you can customize types to have more or less prefix.

Author: For this one, I am still unsure on how to proceed since we do not squash commits. I would like to have the list of PRs with author ordered as shown by changelogithub.

### Submit a Pull Request

If you don't know how to send a Pull Request, we recommend reading [the guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

When sending a pull request, make sure your PR's title also follows the [Commit Convention](#commit-convention).

We follow [trunk based development](https://www.atlassian.com/continuous-delivery/continuous-integration/trunk-based-development) to contribute to our repositories.

The core principles are as follow: (we advise you to read the previous link)
- Developpers merge small and frequent updates to the main branch
- Developpers create short-lived branches with a few small commits
- CI/CD runs on every push and check that tests (unitary, functionnal and integration) and formatting still passes
- Release are produced from the main branch

Thus, a Pull Request (PR) should target the main branch.

The following coding rules must be respected:

- All features or bug fixes __must be tested__ by one or more unit tests.
- All public APIs must be documented.
- Formatting, style and code quality rules must be applied.

> If your PR fixes or resolves an existing issue, please link in the desciption of your PR. You can either put the issue number or the complete link to the issue if the fix is not in the same repository as the issue. You will have to replace the 123 by the issue number you fix.

Reviewer: You can add a pull request template like in GUI https://github.com/aneoconsulting/ArmoniK.Admin.GUI/blob/main/.github/PULL_REQUEST_TEMPLATE.md

Example:
```markdown
fix #123
fix https://github.com/aneoconsulting/ArmoniK/issues/123
```

This will let GitHub know the issues are linked, and automatically close them once the PR gets merged. Learn more at [the guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### Review a Pull Request

Reviewers should ensure that the PR submission guidelines are respected.

We should particularly take attention to:
- Documentation
- Tests
- Maintainability
- Performances

> ## Release

Reviewer: we try to follow semver (https://semver.org) as much as we can

Example:
```markdown
fix #123
fix https://github.com/aneoconsulting/ArmoniK/issues/123
```

This will let GitHub know the issues are linked, and automatically close them once the PR gets merged. Learn more at [the guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### Review a Pull Request

Reviewers should ensure that the PR submission guidelines are respected.

We should particularly take attention to:
- Documentation
- Tests
- Maintainability
- Performances

## Release

Our release process in the repositories is as follow:

1. Create a release and a tag from the GitHub interface. We try to follow [SemVer](https://semver.org/) as much as possible for our tags.
2. The CI/CD will check that the tag is valid. If the tag is not valid, the CI/CD will fail.
3. The CI/CD will run and produce the artifacts with the given tag.

> Note: Some repositories may have a slightly different way of producing releases, therefore refer to the documentation in the repositories to ensure that you follow the right steps. If the process is still not clear for you, do not hesitate to ask questions!

Reviewer 2: ```suggestion
Note: Some repositories may have a slightly different way of producing releases; therefore, refer to the documentation in the repositories to ensure that you follow the right steps. If the process is still not clear to you, do not hesitate to contact us.
```
